src/node_locks.cc

```diff
@@ -0,0 +1,738 @@
+  // stolen.
+  if (!lock->is_stolen()) {
+    if (was_rejected) {
+      // The callback promise was rejected, so reject the final promise
+      lock->released_promise()->Reject(context, callback_result).Check();
+    } else {
+      // The callback promise was fulfilled, so resolve the final promise
+      lock->released_promise()->Resolve(context, callback_result).Check();
+    }
+  }
+
+  ProcessQueue(env);
+}
+
+// Remove a lock from held_locks_ when it's no longer needed
+void LockManager::ReleaseLock(Lock* lock) {
+  const std::u16string& resource_name = lock->name();
+  auto resource_iter = held_locks_.find(resource_name);
+  if (resource_iter == held_locks_.end()) return;
+
+  auto& resource_locks = resource_iter->second;
+  for (auto lock_iter = resource_locks.begin();
+       lock_iter != resource_locks.end();
+       ++lock_iter) {
+    if (lock_iter->get() == lock) {
+      resource_locks.erase(lock_iter);
+      if (resource_locks.empty()) held_locks_.erase(resource_iter);
+      break;
+    }
+  }
+}
+
+// Wakeup of target Environment's event loop
+void LockManager::WakeEnvironment(Environment* target_env) {
+  if (target_env == nullptr || target_env->is_stopping()) return;
+
+  target_env->SetImmediateThreadsafe([](Environment* env_to_wake) {
+    if (env_to_wake != nullptr && !env_to_wake->is_stopping()) {
+      LockManager::GetCurrent()->ProcessQueue(env_to_wake);
+    }
+  });
+}
+
+// Remove all held locks and pending requests that belong to an Environment
+// that is being destroyed
+void LockManager::CleanupEnvironment(Environment* env_to_cleanup) {
+  Mutex::ScopedLock scoped_lock(mutex_);
+
+  // Remove every held lock that belongs to this Environment.
+  for (auto resource_iter = held_locks_.begin();
+       resource_iter != held_locks_.end();) {
+    auto& resource_locks = resource_iter->second;
+    for (auto lock_iter = resource_locks.begin();
+         lock_iter != resource_locks.end();) {
+      if ((*lock_iter)->env() == env_to_cleanup) {
+        lock_iter = resource_locks.erase(lock_iter);
+      } else {
+        ++lock_iter;
+      }
+    }
+    if (resource_locks.empty()) {
+      resource_iter = held_locks_.erase(resource_iter);
+    } else {
+      ++resource_iter;
+    }
+  }
+
+  // Remove every pending request submitted by this Environment.
+  for (auto request_iter = pending_queue_.begin();
+       request_iter != pending_queue_.end();) {
+    if ((*request_iter)->env() == env_to_cleanup) {
+      request_iter = pending_queue_.erase(request_iter);
+    } else {
+      ++request_iter;
+    }
+  }
+
+  // Finally, remove it from registered_envs_
+  registered_envs_.erase(env_to_cleanup);
+}
+
+// Cleanup hook wrapper
+void LockManager::OnEnvironmentCleanup(void* arg) {
+  Environment* env = static_cast<Environment*>(arg);
+  LockManager::GetCurrent()->CleanupEnvironment(env);
+}
+
+static Local<Object> CreateLockInfoObject(Isolate* isolate,
```

Comment:
Consider having this return a `MaybeLocal<Object>` to better facilitate error propagation